File: api/src/models/sending_profile.py
from typing import TYPE_CHECKING, Final, Optional
from sqlmodel import Relationship, SQLModel, Field

from src.models.custom_header import CustomHeaderCreate
from src.models.phishing_kit import PhishingKitSendingProfileLink

if TYPE_CHECKING:
    from src.models.campaign import Campaign
    from src.models.custom_header import CustomHeader
    from src.models.phishing_kit import PhishingKit


class CampaignSendingProfileLink(SQLModel, table=True):
    """Many-to-many link between Campaign and SendingProfile."""
    campaign_id: Optional[int] = Field(
        default=None, foreign_key="campaign.id", primary_key=True
    )
    sending_profile_id: Optional[int] = Field(
        default=None, foreign_key="sendingprofile.id", primary_key=True
    )


class SendingProfile(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    smtp_host: str
    smtp_port: int
    username: str
    password: str
    from_fname: str
    from_lname: str
    from_email: str

    # FK

    realm_name: Optional[str] = Field(
        default=None, foreign_key="realm.name", index=True
    )

    # Relationships

    campaigns: list["Campaign"] = Relationship(
        back_populates="sending_profiles", link_model=CampaignSendingProfileLink
    )
    custom_headers: list["CustomHeader"] = Relationship(
        back_populates="profile",
        sa_relationship_kwargs={"cascade": "all, delete-orphan"},
    )
    phishing_kits: list["PhishingKit"] = Relationship(
        back_populates="sending_profiles", link_model=PhishingKitSendingProfileLink
    )
    realm: Optional["Realm"] = Relationship(back_populates="sending_profiles")


class SendingProfileCreate(SQLModel):
    name: str
    smtp_host: str
    smtp_port: int
    username: str
    password: str
    from_fname: str
    from_lname: str
    from_email: str

    custom_headers: list[CustomHeaderCreate] = []


class SendingProfileDisplayInfo(SQLModel):
    id: int
    name: str
    from_fname: str
    from_lname: str
    from_email: str
    smtp_host: Final[str]
    smtp_port: Final[int]

File: web/src/types/sendingProfile.ts
// 1. A peça pequena (Header)
export interface CustomHeader {
  id?: number; // Opcional no frontend ao criar
  name: string;
  value: string;
}

// 2. A receita para criar (sem ID)
export interface SendingProfileCreate {
  name: string;
  smtp_host: string;
  smtp_port: number;
  username: string;
  password: string;
  from_fname: string;
  from_lname: string;
  from_email: string;
  custom_headers?: { name: string; value: string }[];
}

// 3. O produto final que vem da BD (com ID)
export interface SendingProfile extends SendingProfileCreate {
  id: number;
  realm_name?: string;
  custom_headers?: (CustomHeader & { id: number })[];
}

// 4. Subset returned by GET /sending-profiles (list endpoint)
//    Matches backend SendingProfileDisplayInfo — no password/username exposed
export interface SendingProfileDisplayInfo {
  readonly id: number;
  readonly name: string;
  readonly from_fname: string;
  readonly from_lname: string;
  readonly from_email: string;
  readonly smtp_host: string;
  readonly smtp_port: number;
}
